### DISCUSSION TITLE
Search in ID's

### DISCUSSION THREAD

**Discussion by Sushanti99** [milvus-io/milvus#4731] [Q&A and General discussion] [answered]
Here's a question:

Can I narrow down the vector search in milvus's collection? Suppose I have ID's of [1,2,3] after inserting in milvus. Can I conduct a search on milvus only looking for ID's [1,2] and get the distance?

**Comment by shiyu22** [comment]
Hi, your question sounds interesting, and can you tell me what scenario do you need to search in ids?

Actually, it doesn't have this feature, but if you have a requirement you can submit a [feature issue](https://github.com/milvus-io/milvus/issues) to us.

**Reply by Sushanti99** [reply]
Elaborating the feature requirement here:

Creating partitions is a good idea in Milvus but there's a performance limitation of 4096 number of collections and partitions that can be created in one Milvus cluster. As far as I know, the performance decreases exponentially after this number.

I want to be able to store vectors in billions in one Milvus cluster - so I only create 1 collection in a Milvus cluster.
When I want to combine this with search, the ID's returned by Milvus are from the entire collection. But that's not what I'm looking for. I have a list of ID's that I want to search on - say [1,2] but the collection has vectors [1,2,3,4]. Instead of searching and returning distance metrics for the entire collection's vectors ([1,2,3,4]), I want to only search and rank on the basis of [1,2]. 

Currently I'm doing this in naive approach where I retrieve search results, then compare from the top of the ID's [1,2] and rank them myself but I was wondering if there's a better/easier way of doing this.

**Reply by Sushanti99** [reply]
I also got to wondering if https://milvus.io/docs/v0.11.0/search_entity.md#Python this would be helpful. specify in the DSL under terms which ID's I'm looking for or something but I could not experiment around with this because this is Milvus 11.0 and there's no pymilvus support for that. 

Asking the community for the best solution to this. Also willing to contribute if there are discussions already happening around/of this topic.

**Comment by yhmo** [comment]
This requirement is similar to #4667
In the #4667, user get the filtered IDs coming from Solr and don't want to search the whole collection of the milvus db, just need a subset of the collection by specified IDs.
Currently milvus 0.10.x has no ability to allow user specify IDs to search. We did some experiment in 0.11.x(stop maintain now) to filter vector by given condition, it can meet this requirement. We are now design/develop a new architect cloud-native milvus, all the experience will be bring into the new milvus. And the new milvus will be also open-source, too.

**Comment by yhmo** [comment]
Partition can narrow your search scope, but huge partition number can bring overhead to meta system.
A possible solution is to break the limitation of partition number:
1. add the following lines to server_config.yaml
```
engine:
  max_partition_num: 32768 
```

2. restart milvus server

The "max_partition_num" is a hidden config item. After this change, you can create at most 32768 partitions in milvus. You also can set a larger number by the "max_partition_num".
Then arrange your vectors by IDs, and insert them into partitions by range, for example:
You have vectors: v1, v2, v3.... , v10000
The vectors has ids: 1, 2, 3, .... 10000
You create 100 partitions: p1, p2, ... p100
Insert by range: v1 ~ v100 insert into p1, v101 ~ v200 insert into p2, ..... v9900 ~ v10000 insert into p100

Now you have IDs: v50, v99, v140 to search
You know they are in p1 and p2, so you only search p1 and p2. But the top_k need to set to 100 to ensure the v50, v99, v140 is in result.

**Reply by Sushanti99** [reply]
I can't use the upper count of partitions. I will need to create partitions as and when required and need to be able to search in them. 

Similar to elasticsearch, if I have a variable called ObjectID - it's associated with subgroups under it and I can search in that. But the issue with Milvus's design is that something like that cannot be supported/scaled in this scenario. I've also mentioned the same in https://github.com/milvus-io/milvus/issues/4734

**Comment by Sushanti99** [comment]
There are 2 ways I see it. 
1. Go on creating partitions and adding more milvus servers as and when needed. This significantly reduces the efficiency because considering Milvus can store vectors in billions scale, if I store probably 2000 vectors in each partition and I'm only to create only 32768 partitions, that's not using resources well
2. I insert vectors in the same milvus server, in billions as and when we go. And search on it based on ID's - because this is not supported currently, I'm doing this very inefficiently. 

If we can make the second solution happen in an efficient manner, that would be a good design.

**Comment by yhmo** [comment, marked as answer]
@Sushanti99 
You can create a collection for each video.
We didn't limit the collection number.
I do some test on my machine, create 20000 collections, each collection insert 2000 vectors. It totally cost 76GB disk storage.
Create 20000 collection time cost: 55 minutes
Search(nq=1, topk=10) in one collection, time cost: 30~50ms

The script:

```
import time
import numpy as np
from milvus import Milvus, IndexType, Status
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'
TABLE_DIMENSION = 128
PRINT_VECTOR = False

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)

def gen_vec_list(nb, seed=np.random.RandomState(1234)):
    xb = seed.rand(nb, TABLE_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    if PRINT_VECTOR == True:
        print(vec_list)
    return vec_list

vec_list = gen_vec_list(2000, seed=np.random.RandomState(10))
def add(tb_name):
    milvus = Milvus(host=SERVER_ADDR, port=SERVER_PORT)
    status, result = milvus.insert(collection_name=tb_name, records=vec_list)
    status = milvus.flush(collection_name=tb_name)

def insert():
    time_start = time.time()
    for k in range(20000):
        print("No.", k)


        tb_name = 'tbl_' + str(k)
        tb_param =  {'collection_name': tb_name, 'dimension': TABLE_DIMENSION, 'index_file_size': 4, 'metric_type':MetricType.L2}
        status = MILVUS.create_collection(tb_param)

        add(tb_name)
        time_end = time.time()
        print("done", time_end - time_start, 'sec')

def search(table_name):
    SEARCH_PARAM = {'nprobe': 256}
    query_vectors = gen_vec_list(nb=1, seed=np.random.RandomState(1))
    time_start = time.time()
    status, results = MILVUS.search(collection_name=table_name, query_records=query_vectors,
                                    top_k=10, params=SEARCH_PARAM)
    time_end = time.time()
    print("search time cost:", time_end - time_start, 'sec')

if __name__ == "__main__":
    insert()
    for k in range(100):
        print("No.", k)
        tb_name = 'tbl_' + str(k)
        search(tb_name)
```

**Reply by Sushanti99** [reply]
Thank you for this. Will do some performance analysis of my own too. This is really helpful

**Reply by yhmo** [reply]
"Search(nq=1, topk=10) in one collection, time cost: 30 ~ 50ms"
After investigating the time cost of each phase, I found the SQL cost become bottleneck with huge collection number. 
So far I found three impacts caused by huge number of collections:
1. Before query, milvus performs SQL statements to MySQL to get collection and segments information. There are two SQL statements executed, they cost 20 ~ 25ms. Compute vector distance in each segment only cost 5ms.

2. When restart milvus, the wal module collect all collections information to check wal flag, it performs SQL statement for each collection, totally cost 5 minutes to start the server.

3. All the data(20000 collections, 2000 vector per collection) is about 76GB, including 20GB(2000 files) raw vectors data and 56GB(2000 files) bloom filter data. Each bloom filter file is 2.8MB(for estimation of each segment has 500000 vectors). With collection number increase, bloom filter files occupy too much disk storage.
To make all raw data can be keep in cache, just modify the cache.cache_size to be greater than 20GB:
```
cache:
  cache_size: 25GB
```

**Comment by yhmo** [comment]
My test statistics:

- 100 collections, each collection insert 2000 vectors
  create collection time cost: 10 seconds
  search time cost: 5 ~ 20ms

- 5000 collections, each collection insert 2000 vectors
  create collection time cost: 12 minutes
  search time cost: 10 ~ 30ms

- 20000 collections, each collection insert 2000 vectors
  create collection time cost: 56 mibutes
  search time cost: 30 ~ 50ms

**Reply by tianya092** [reply]
请问搜索时间成本，是单个partition里搜索的成本吗？我现在的按天分区，有时需要在单个partition搜索，有时需要在整个集合搜索，但是分区多了之后（4000），整个集合搜索单次查询需要4s,太慢了, 不知道有啥好的建议？

**Reply by yhmo** [reply]
在单个partition搜索时，4000分区数量对性能影响不大
但在整个集合搜索时，分区数量越多，性能越差，而且影响比较明显。
我们限制分区数量在4096是为了避免用户滥用分区，导致性能慢到无法接受。没啥好的方法可以避免，只能建议尽量减少分区数量。

**Reply by tianya092** [reply]
如果不按天分区的话，那么检索出来的，可能不是我需要的日期范围的向量，这个很矛盾。而且通过调整nlist和nprobe参数，几乎没有任何性能提升。同时我测试了一下同样体量的数据，不分区，全量查询的时间仅需22ms

**Reply by yhmo** [reply]
全表搜索之前要访问mysql/sqlite的。先是用sql语句找出collection里有多少个partition，然后再对每个partition使用sql语句找出每个partition里有哪些segment.
如果只有1个partition，那么后面那条sql只需执行1次。如果有4000个，那么后面那条sql要执行很多次，虽然做了一些优化，对sql做了合并，但仍然会有多次，所以瓶颈都在执行sql上了。
